clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -128,23 +137,99 @@ record_filter_t::parallel_shard_error(void *shard_data)
+    ++shard->output_entry_count;
+    return true;
+}
+
 bool
 record_filter_t::parallel_shard_memref(void *shard_data, const trace_entry_t &entry)
 {
     per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+    ++per_shard->input_entry_count;
     bool output = true;
+    bool all_stop = true;
     for (int i = 0; i < (int)filters_.size(); ++i) {
-        if (!filters_[i]->parallel_shard_filter(entry, filter_shard_data_[i]))
+        bool stop = false;
+        if (!filters_[i]->parallel_shard_filter(entry, per_shard->filter_shard_data[i],
+                                                &stop))
             output = false;
+        if (stop)
+            per_shard->filter_stop[i] = true;
+        all_stop = all_stop && per_shard->filter_stop[i];
+    }
+
+    // Everything before the timestamp marker in the first unit header is part of the
+    // trace shard's header.
+    if (entry.type == TRACE_TYPE_MARKER && entry.size == TRACE_MARKER_TYPE_TIMESTAMP) {
+        per_shard->in_shard_header = false;
+    }
+
+    // We output everything in the trace shard's header.
+    if (per_shard->in_shard_header) {
+        output = true;
+    }
+
+    // Keep track of the last omitted unit header so that we can output it if and when
+    // we output a trace_entry_t from this unit.
+    if (!output && entry.type == TRACE_TYPE_MARKER) {
+        switch (entry.size) {
+        case TRACE_MARKER_TYPE_TIMESTAMP:
+            // No need to remember the previous unit's header anymore. We're in the
+            // next unit now.
+            per_shard->last_filtered_unit_header.clear();
+            ANNOTATE_FALLTHROUGH;
+        case TRACE_MARKER_TYPE_WINDOW_ID:
+        case TRACE_MARKER_TYPE_CPU_ID:
+            per_shard->last_filtered_unit_header.push_back(entry);
+        }
+    }
+
+    // Since we're outputing something from this unit, output its header.
+    if (output) {
+        for (trace_entry_t &unit_header_entry : per_shard->last_filtered_unit_header) {
+            if (!write_trace_entry(per_shard, unit_header_entry))
+                return false;
+        }
+        per_shard->last_filtered_unit_header.clear();
+    }
+
+    if (!output) {
+        if (entry.type == TRACE_TYPE_FOOTER ||
+            (entry.type == TRACE_TYPE_MARKER &&
+             entry.size == TRACE_MARKER_TYPE_CHUNK_FOOTER)) {
+            output = true;
+        } else if (!all_stop && entry.type == TRACE_TYPE_MARKER) {
+            // Markers that may be required for future instr/memrefs, therefore
+            // must be retained if there's a chance we'll output more entries
+            // later. These are not really tied to a unit header, so to save
+            // space we do not output their unit's header unless some other
+            // instr/memref is output from the same unit. We decided against
+            // buffering them since there may be too many of these.
+            // XXX: This list needs to be kept in sync with our trace format.
```

Comment:
I'm not sure I understand why we are going to all this work?  What breaks down if we just let the tool remove what it wants?  If it removes all the non-header entries in one unit it seems fine to leave headers with nothing else there.  If it wants to remove markers we should let it; if that breaks the trace it's the tool writer's fault.  It seems unlikely a typical tool would hit problems: isn't it going to be removing cache-filtered loads or encoding entries and nothing else?  And those cases do not need any of this logic?

Reply:
> What breaks down if we just let the tool remove what it wants?

This was motivated by the toggle_filter where we want to remove everything before or after a certain instr count. Of course, doing just that is not acceptable since it might remove necessary header/footer entries (which is a clear error), or even encoding, v2p entries (which may silently break some stuff). So some of these shouldn't be removed. I ended up generalising these rules to all filters, instead of just the toggle_filter.

About the "let the tool be more responsible" approach: one drawback was that the tool needs to be aware of all internal entries and remember to not filter them out. But, rethinking now, that's what the record_analysis_tools sign up for, so maybe that's fine.

Perhaps this logic should be moved to inside  the toggle_filter instead of the record_filter tool.

> If it removes all the non-header entries in one unit it seems fine to leave headers with nothing else there.

This was to optimize space, which is the primary goal for the current filtering use cases.